### ISSUE TITLE
Needlessly building ":build" deps for formulas that are already installed

### ISSUE THREAD

**Issue by chdiza** [Homebrew/brew#1272]
[Transferred from Homebrew/homebrew-core#5740 at @MikeMcQuaid's request]

- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?

I noticed this problem with `diff-pdf`, but it's probably because of a bug in `brew`'s deciding which deps need to be installed.

`brew install --build-from-source diff-pdf` will incorrectly try to build and install `:build` deps for `openjpeg`, even when `openjpeg` has already been poured from a bottle.

Reproduce: Pour the bottles for `diff-pdf`'s dependencies: `brew install wxmac cairo poppler`  Notice that `openjpeg` is now installed, since it's a recursive dep.  Other recursive deps are poured too.

Now, do `brew install --build-from-source diff-pdf`.  What should happen is that `diff-pdf` starts being built forthwith (all of its deps are already installed).  What actually happens is that `brew` starts trying to build `cmake`, `doxygen`, and `sphinx-doc`, because those are deps of `openjpeg`.  But, those are *:build* deps of `openjpeg`, and `openjpeg` has already been poured and is therefore already built.

My *guess* is that `brew` is simply failing to filter out the ":build" deps from its list of things it needs to do before it can build `diff-pdf`.  Even though `diff-pdf` is going to be built rather than poured, that doesn't mean that all of its deps need to be :built.

**Comment by apjanke** [comment]
I can reproduce.

```
$ brew info diff-pdf | grep Build:
Build: pkg-config, automake, autoconf
$ brew info openjpeg | grep Build:
Build: cmake, doxygen
$ brew install -s diff-pdf
==> Installing dependencies for diff-pdf: cmake, doxygen
==> Installing diff-pdf dependency: cmake
```

I think your analysis is basically right, and it's not filtering the `:build` deps out. The `--build-from-source` flag is probably contaminating the dependency expansion, probably [somewhere around here in `formula_installer.rb`](https://github.com/Homebrew/brew/blob/de880f1e870d91dd838228de2987f9bd6d194aef/Library/Homebrew/formula_installer.rb#L384-L404). (We used to have an issue where the build from source option was basically treated as global, so dependencies would also be built from source. That was fixed, but this is probably a related leftover from that logic.)

Doing the filtering is going to be a little tricky, because that dependency expansion is implemented as a recursive expansion where your pruning logic only sees one level of expansion at a time and doesn't know what level it's at, and the "build from source" is represented as just a command line argument, not part of the `BuildOptions` object. The filtering logic should be something like "prune all `:build` deps when at level > 0". Might be time to do some rearranging in there. Or we could just hack it by testing whether `dependent` is in the list of formulae on the command line.

I'm thinking the Right Way to handle this is probably to add `--build-from-source` to `BuildOptions`.

**Comment by apjanke** [comment]
Well, `BuildOptions` grabs flags from `ARGV` in the constructor using `ARGV`'s `flags` logic, so it would be a pretty extensive operation to change it to represent whether formulae should be built from source.

But it looks like `ARGV.build_formula_from_source?` could be used inside the dependency-expansion logic to do the pruning of unnecessary build-time dependencies. I'll put in a PR.

**Comment by apjanke** [comment]
PR to fix: #1305
